### PR TITLE
multiplex module

### PR DESCRIPTION
**Pull request by ssryps**
multiplex module( 2019 os training): 
1. modify poll, select implementation
2. add epoll

### CODE REVIEW

File: kernel/src/trap.rs [hunk 1]
```diff
@@ -2,23 +2,34 @@ use crate::arch::cpu;
...
 pub fn uptime_msec() -> usize {
-    unsafe { crate::trap::TICK / crate::consts::USEC_PER_TICK / 1000 }
+    unsafe { crate::trap::TICK * crate::consts::USEC_PER_TICK / 1000 }
 }
 
 pub fn timer() {
     if cpu::id() == 0 {
         unsafe {
             TICK += 1;
+            if uptime_msec() % 50 == 0 {
```
Comment (jiegec) [inline kernel/src/trap.rs]: Move 50 to const

File: kernel/src/fs/file_like.rs [hunk 2]
```diff
@@ -41,6 +51,9 @@ impl FileLike {
                     FileLike::Socket(socket) => {
                         socket.ioctl(request, arg1, arg2, arg3)?;
                     }
+                    FileLike::EpollInstance(instance) => {
+						return Err(SysError::ENOSYS);
```
Comment (equation314) [inline kernel/src/fs/file_like.rs]: Indent using spaces

File: kernel/src/syscall/fs.rs [hunk 3]
```diff
@@ -172,20 +195,25 @@ impl Syscall<'_> {
...
+        Condvar::wait_events(condvars.as_slice(), move || {
             let proc = self.process();
             let mut events = 0;
             for (&fd, file_like) in proc.files.iter() {
-                if fd >= nfds {
-                    continue;
-                }
+//                if fd >= nfds {
+//                    continue;
+//                }
```
[the remark this answers is not in the document]
Comment (ssryps) [reply kernel/src/syscall/fs.rs]: files is a btreemap, so fd is not an index but a key. nfds is the size of input fdset. It's meaningless to compare these two numbers.

File: kernel/src/syscall/fs.rs [hunk 4]
```diff
@@ -228,6 +256,252 @@ impl Syscall<'_> {
...
+        if !proc.pid.is_init() {
+            // we trust pid 0 process
+            info!("sys_epoll_ctl: epfd: {}, op: {:?}, fd: {:#x}", epfd, op, fd);
+        }
+
+        let slice = unsafe { self.vm().check_read_array(event, 1)? };
+
+        if proc.files.get(&fd).is_none(){
+            return Err(SysError::EPERM);
+        }
```
[the remark this answers is not in the document]
Comment (ssryps) [reply kernel/src/syscall/fs.rs]: the parameter fd may not be valid, so the check is needed.

File: kernel/src/syscall/fs.rs [hunk 5]
```diff
@@ -228,6 +256,252 @@ impl Syscall<'_> {
...
+
+        let epollInstance = match proc.get_epoll_instance(epfd) {
+            Ok(ins) => ins,
+            Err(err) => {
+                return Err(err);
+            }
+        };
```
[the remark this answers is not in the document]
Comment (ssryps) [reply kernel/src/syscall/fs.rs]: different err type

[the remark this answers is not in the document]
Comment (wangrunji0408) [reply kernel/src/syscall/fs.rs]: you can `map_err`

File: kernel/src/syscall/fs.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply kernel/src/syscall/fs.rs]: sorry I got it wrong

**Pull request merged**